File: scripts/extract_allocations.py
#!/usr/bin/env python3
import argparse
import json
import pathlib
import re

import pdfplumber


TARGET_HEADING = "在各个国家（地区）证券市场的股票及存托凭证投资分布"
NUMBER = re.compile(r"(-?\d+(?:\.\d+)?)\s*$")
COUNTRY_MAP = {
    "美国": "美国", "中国": "中国", "中国大陆": "中国", "中国内地": "中国",
    "中国香港": "中国香港", "香港": "中国香港",
    "韩国": "韩国", "日本": "日本",
}


def clean_country(value):
    return re.sub(r"\s+", "", value).strip("：:")


def parse_table(text):
    compact = text.replace("(", "（").replace(")", "）")
    if TARGET_HEADING not in compact:
        return None
    lines = compact.splitlines()
    start = next(i for i, line in enumerate(lines) if TARGET_HEADING in line)
    rows = []
    total = None
    footnote = ""
    for line in lines[start + 1:]:
        stripped = line.strip()
        if stripped.startswith("注：") and rows:
            footnote = stripped
            break
        match = NUMBER.search(stripped)
        if not match:
            continue
        pct = float(match.group(1))
        prefix = stripped[:match.start()].strip()
        if prefix.startswith("合计"):
            total = pct
            break
        country = clean_country(prefix.split()[0] if prefix.split() else "")
        if not country or country in {"国家（地区）", "公允价值（人民币元）"}:
            continue
        if any(token in country for token in ["占基金", "资产净值", "比例"]):
            continue
        rows.append((country, pct))
    if not rows or total is None:
        return None
    return rows, total, footnote


def extract_one(item):
    found = None
    with pdfplumber.open(item["pdf_path"]) as pdf:
        page_texts = [page.extract_text(x_tolerance=2, y_tolerance=2) or "" for page in pdf.pages]
        for page_index, text in enumerate(page_texts):
            # Country tables occasionally continue with the total and footnote
            # on the following page.
            window = text + "\n" + (page_texts[page_index + 1] if page_index + 1 < len(page_texts) else "")
            parsed = parse_table(window)
            if parsed:
                rows, total, footnote = parsed
                found = (page_index + 1, rows, total, footnote)
                break
    if not found:
        raise RuntimeError(f"{item['code']} {item['short_name']}: country table not found")
    page_number, rows, reported_total, footnote = found
    allocation = {name: 0.0 for name in ["美国", "中国", "中国香港", "韩国", "日本", "其他"]}
    raw_rows = []
    for country, pct in rows:
        category = COUNTRY_MAP.get(country, "其他")
        allocation[category] += pct
        raw_rows.append({"country": country, "category": category, "percent_nav": pct})
    calculated_total = sum(allocation.values())
    residual = 100.0 - reported_total
    allocation["现金及其他"] = residual
    return {
        **{key: value for key, value in item.items() if key != "pdf_path"},
        "report_date": item["period"][:4] + "-" + {"Q1":"03-31", "Q2":"06-30", "Q3":"09-30", "Q4":"12-31"}[item["period"][4:]],
        "source_page": page_number,
        "classification_note": footnote,
        "raw_country_rows": raw_rows,
        "reported_equity_country_total": reported_total,
        "calculated_country_total": round(calculated_total, 6),
        "allocation": {key: round(value, 6) for key, value in allocation.items()},
    }


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--manifest", required=True)
    parser.add_argument("--output", required=True)
    args = parser.parse_args()
    manifest = json.loads(pathlib.Path(args.manifest).read_text(encoding="utf-8"))
    result = [extract_one(item) for item in manifest]
    pathlib.Path(args.output).write_text(json.dumps(result, ensure_ascii=False, indent=2), encoding="utf-8")
    print(f"Extracted {len(result)} reports")


if __name__ == "__main__":
    main()

File: scripts/extract-allocations.mjs
#!/usr/bin/env node
import { readFile } from "node:fs/promises";
import path from "node:path";
import { fileURLToPath, pathToFileURL } from "node:url";

import { getDocument } from "pdfjs-dist/legacy/build/pdf.mjs";

import { parseArgs, requireArgs } from "./lib/cli.mjs";
import { readJson, writeJson } from "./lib/files.mjs";

const TARGET_HEADING = "在各个国家（地区）证券市场的股票及存托凭证投资分布";
const CATEGORIES = ["美国", "中国", "中国香港", "韩国", "日本", "其他"];
const COUNTRY_MAP = new Map([
  ["美国", "美国"], ["中国", "中国"], ["中国大陆", "中国"], ["中国内地", "中国"],
  ["中国香港", "中国香港"], ["香港", "中国香港"], ["韩国", "韩国"], ["日本", "日本"],
]);
const STANDARD_FONT_DATA_URL = `${fileURLToPath(new URL("../node_modules/pdfjs-dist/standard_fonts/", import.meta.url))}${path.sep}`;

function pageLines(items) {
  const lines = [];
  const sorted = items
    .filter((item) => typeof item.str === "string" && item.str.length > 0)
    .map((item) => ({ text: item.str, x: item.transform[4], y: item.transform[5], width: item.width ?? 0 }))
    .sort((left, right) => Math.abs(right.y - left.y) > 2 ? right.y - left.y : left.x - right.x);
  for (const item of sorted) {
    let line = lines.find((candidate) => Math.abs(candidate.y - item.y) <= 2);
    if (!line) {
      line = { y: item.y, parts: [] };
      lines.push(line);
    }
    line.parts.push(item);
  }
  return lines
    .sort((left, right) => right.y - left.y)
    .map((line) => {
      const parts = line.parts.sort((left, right) => left.x - right.x);
      let value = "";
      let previous = null;
      for (const part of parts) {
        if (previous && !/\s$/.test(value) && !/^\s/.test(part.text)) {
          const gap = part.x - (previous.x + previous.width);
          if (gap > 3) value += " ";
        }
        value += part.text;
        previous = part;
      }
      return value.replace(/\s+/g, " ").trim();
    });
}

function cleanCountry(value) {
  return value.replace(/\s+/g, "").replace(/^[：:]+|[：:]+$/g, "");
}

export function parseTable(text) {
  const lines = text.replaceAll("(", "（").replaceAll(")", "）").split(/\r?\n/);
  const start = lines.findIndex((line) => line.replace(/\s+/g, "").includes(TARGET_HEADING));
  if (start < 0) return null;
  const rows = [];
  let total = null;
  let footnote = "";
  for (const rawLine of lines.slice(start + 1)) {
    const line = rawLine.trim();
    if (line.replace(/\s+/g, "").startsWith("注：") && rows.length) {
      footnote = line;
      break;
    }
    const match = /(-?\d+(?:\.\d+)?)\s*$/.exec(line);
    if (!match) continue;
    const percent = Number(match[1]);
    const prefix = line.slice(0, match.index).trim();
    if (prefix.replace(/\s+/g, "").startsWith("合计")) {
      total = percent;
      break;
    }
    const country = cleanCountry(prefix.split(/\s+/)[0] ?? "");
    if (!country || ["国家（地区）", "公允价值（人民币元）"].includes(country)) continue;
    if (["占基金", "资产净值", "比例"].some((token) => country.includes(token))) continue;
    rows.push([country, percent]);
  }
  return rows.length && total !== null ? { rows, total, footnote } : null;
}

async function extractPageTexts(pdfPath) {
  const data = new Uint8Array(await readFile(pdfPath));
  const pdf = await getDocument({ data, disableWorker: true, standardFontDataUrl: STANDARD_FONT_DATA_URL }).promise;
  const texts = [];
  for (let pageNumber = 1; pageNumber <= pdf.numPages; pageNumber += 1) {
    const page = await pdf.getPage(pageNumber);
    const content = await page.getTextContent();
    texts.push(pageLines(content.items).join("\n"));
  }
  await pdf.destroy();
  return texts;
}

async function extractOne(item) {
  const texts = await extractPageTexts(item.pdf_path);
  let found = null;
  for (let index = 0; index < texts.length; index += 1) {
    const window = `${texts[index]}\n${texts[index + 1] ?? ""}`;
    const parsed = parseTable(window);
    if (parsed) {
      found = { ...parsed, page: index + 1 };
      break;
    }
  }
  if (!found) throw new Error(`${item.code} ${item.short_name}: country table not found`);
  const allocation = Object.fromEntries(CATEGORIES.map((category) => [category, 0]));
  const rawRows = found.rows.map(([country, percent]) => {
    const category = COUNTRY_MAP.get(country) ?? "其他";
    allocation[category] += percent;
    return { country, category, percent_nav: percent };
  });
  const calculatedTotal = Object.values(allocation).reduce((sum, value) => sum + value, 0);
  allocation["现金及其他"] = 100 - found.total;
  const { pdf_path: ignoredPdfPath, ...evidence } = item;
  const quarterDates = { Q1: "03-31", Q2: "06-30", Q3: "09-30", Q4: "12-31" };
  return {
    ...evidence,
    report_date: `${item.period.slice(0, 4)}-${quarterDates[item.period.slice(4)]}`,
    source_page: found.page,
    classification_note: found.footnote,
    raw_country_rows: rawRows,
    reported_equity_country_total: found.total,
    calculated_country_total: Number(calculatedTotal.toFixed(6)),
    allocation: Object.fromEntries(Object.entries(allocation).map(([key, value]) => [key, Number(value.toFixed(6))])),
  };
}

export async function main(argv = process.argv.slice(2)) {
  const args = parseArgs(argv);
  requireArgs(args, ["manifest", "output"]);
  const manifest = await readJson(args.manifest);
  const result = [];
  for (const item of manifest) result.push(await extractOne(item));
  await writeJson(args.output, result);
  console.log(`Extracted ${result.length} reports`);
}

if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {
  main().catch((error) => {
    console.error(error.stack ?? error.message);
    process.exitCode = 1;
  });
}
